packages/endo/src/archive.js

```diff
@@ -0,0 +1,158 @@
+/* global StaticModuleRecord */
+/* eslint no-shadow: 0 */
+
+import { writeZip } from "./zip.js";
+import { resolve, join } from "./node-module-specifier.js";
+import { compartmentMapForNodeModules } from "./compartmap.js";
+import { search } from "./search.js";
+import { assemble } from "./assemble.js";
+
+const { entries, fromEntries } = Object;
+
+const q = JSON.stringify;
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+const makeRecordingImportHookMaker = (
+  read,
+  baseLocation,
+  manifest,
+  errors
+) => packageLocation => {
+  packageLocation = new URL(packageLocation, baseLocation).toString();
```

Comment:
It might be good to make this use a `resolve` helper, like `import.js` did, for parallelism, rather than URL. Or make `import.js` use URL instead of `resolve`.

Reply:
The domain of locations is always URL’s. The module specifier math requires a hook because there are multiple valid maths. In practice, Endo.js might only ever use the Node.js resolver math, but the open web uses a different one. The `Compartment` API can ideally be used to host either, maybe even both coherently.